### ISSUE TITLE
[module] extractors.py parser extractor contract audit

### ISSUE THREAD

**Issue by Tahjali11** [Tahjali11/Mythic-Edge#9]
﻿## Agent Docs

Use:
- `docs/agent_constitution.md`
- `docs/agent_threads/problem_representation.md`
- `docs/agent_threads/module_contract.md`
- `docs/agent_threads/implementation.md`
- `docs/agent_threads/module_fixer.md`
- `docs/agent_threads/review.md`
- `docs/agent_threads/contract_test.md`
- `docs/agent_threads/module_submitter.md`
- `docs/templates/problem_representation.md`
- `docs/templates/workflow_handoff.md`

Tracker: https://github.com/Tahjali11/Mythic-Edge/issues/5
Integration branch: `codex/parser-module-audit-suite`

## Risk Tier

High

## Summary

Audit `src/mythic_edge_parser/app/extractors.py` as the next parser module after `models.py` and `state.py` to make its event-payload extraction, normalization, identity, hand/GRP lookup, and game-result scope contracts explicit.

This is a workflow audit and contract-writing problem representation, not a known bug report.

## What the code is supposed to do

`extractors.py` should provide safe, deterministic extraction helpers that convert parsed MTGA event payload shapes into normalized primitive facts for parser-owned state and downstream parser tools.

It should safely provide:

- local player selection from match-state player lists
- safe dictionary/list coercion for malformed or partial payloads
- top-level, raw, and queued game-state section access
- match ID, game number, turn number, active player, phase, step, and stage extraction
- local team detection from client actions and game-state payloads
- starting player detection from client actions and turn-one game-state payloads
- local private hand instance ID extraction
- instance ID to GRP ID lookup extraction
- game/match scope label normalization
- game-result identity and nested scope-result extraction
- match-scope result detection
- safe event timestamp conversion to ISO text

## What it is actually doing

Current behavior is implemented and partially covered by focused tests, but the intended contract is not yet written in a dedicated module contract.

The workflow should determine whether the current implementation and tests fully describe and protect:

- `_safe_local_player()`
- `_first_present()`
- `_maybe_int()`
- `_safe_dict()` and `_safe_list()`
- raw and queued game-state payload helpers
- `_hydrate_game_state_identity()`
- `_extract_turn_info()`
- `_extract_starting_player_from_client_action()`
- `_extract_local_team_from_client_action()`
- `_extract_local_team_from_game_state()`
- `_extract_starting_player_from_game_state()`
- `_extract_local_private_hand_instance_ids()`
- `_extract_instance_grp_lookup()`
- `_infer_scope_label()`
- `_extract_game_result_scope_result()`
- `_extract_game_result_identity()`
- `_has_match_scope_result()`
- `_event_datetime()` and `_safe_iso()`

## Primary project layer

parser and state interpretation

## First bad value or first place to inspect

No known bad value yet. Start with this inspection order:

1. Safe normalization primitives: `_first_present()`, `_maybe_int()`, `_safe_dict()`, `_safe_list()`, and `_safe_local_player()`.
2. Raw and queued game-state accessors: `_raw_game_state_payload()`, `_raw_game_state_message()`, `_queued_game_state_message()`, `_game_state_dict_section()`, `_game_state_list_section()`, and `_queued_game_info()`.
3. Game-state identity hydration: `_game_state_identity_payload()`, `_game_state_game_info()`, `_game_state_turn_info_payload()`, `_game_state_system_seat_ids()`, `_hydrate_game_state_identity()`, and `_extract_turn_info()`.
4. Player/team helpers: `_player_seat_id()`, `_player_team_id()`, `_extract_local_team_from_game_state()`, and `_extract_starting_player_from_game_state()`.
5. Client-action helpers: `_extract_starting_player_from_client_action()` and `_extract_local_team_from_client_action()`.
6. Hand and card identity helpers: `_extract_local_private_hand_instance_ids()` and `_extract_instance_grp_lookup()`.
7. Game-result helpers: `_infer_scope_label()`, `_is_known_winner()`, `_result_winner()`, `_extract_game_result_scope_result()`, `_extract_game_result_identity()`, and `_has_match_scope_result()`.
8. Timestamp helpers: `_event_datetime()` and `_safe_iso()`.
9. Call sites in `src/mythic_edge_parser/app/state.py`, `src/mythic_edge_parser/app/transforms.py`, `src/mythic_edge_parser/app/runtime_surfaces.py`, `src/mythic_edge_parser/app/arena_id_validation.py`, and `src/mythic_edge_parser/app/grp_id_candidates.py`.
10. Focused tests in `tests/test_app_extractors.py`, plus integration coverage in `tests/test_state.py`, `tests/test_match_summary_from_match_state.py`, `tests/test_transforms.py`, and `tests/test_runtime_surfaces.py`.

## Expected output

Produce and then test these workflow artifacts:

1. `docs/contracts/parser_extractors.md`
   - written by the Module Contract Writer thread
   - based on current `extractors.py` behavior, call sites, and focused tests
   - separates observed behavior, required guarantees, unknowns, and suspected gaps

2. `docs/implementation_handoffs/parser_extractors_comparison.md`
   - written by the Module Implementer thread
   - compares current code to the contract
   - includes fixes only if the contract reveals clear mismatches or missing safeguards

3. `docs/contract_test_reports/parser_extractors.md`
   - written by the Module Reviewer thread in contract-test mode
   - verifies implementation behavior against the contract
   - lists confirmed matches, mismatches, missing tests, and drift classification

## Scope

In scope:

- `src/mythic_edge_parser/app/extractors.py`
- focused tests for extractor behavior
- parser-owned event payload normalization
- raw, queued, and enriched game-state payload shapes
- client action extraction for local team and starting player
- local private hand and instance/GRP lookup extraction
- game result identity and scope extraction
- timestamp extraction behavior
- compatibility with current parser call sites

Out of scope unless separately approved:

- changing workbook schema
- changing webhook payload shape
- deploying Apps Script
- changing row field names in `models.py`
- changing state ownership in `state.py`
- changing raw MTGA parser modules under `src/mythic_edge_parser/parsers/`
- changing gameplay-action semantics outside extractor call-site compatibility
- changing secrets, environment variables, or live workbook state
- deleting debug, archive, helper, runtime, or observability layers

## Risks and likely breakpoints

- Extractor helpers are underscore-prefixed but widely used as shared parser interfaces; the contract should explicitly name which helpers are supported compatibility surfaces.
- Raw game-state and queued game-state payloads can be partial; fallback order must be explicit.
- Local seat, local team, starting player, and active player can be represented with different key names and number/string forms.
- Hand extraction depends on private hand zones and local system seat IDs; incorrect fallback can capture the wrong player's hand.
- GRP lookup extraction must not confuse `grpId`, `grp_id`, `overlayGrpId`, and `overlay_grp_id` behavior.
- Game-result helpers must distinguish game-scope results from match-scope results and avoid promoting match winners into game winners.
- Timestamp fallback to current time is convenient but potentially non-deterministic; the contract should identify where that behavior is acceptable.

## Validation evidence needed

Focused extractor checks:

```powershell
py -m pytest -q tests/test_app_extractors.py
```

Relevant parser-state and transform checks when extractor behavior changes:

```powershell
py -m pytest -q tests/test_state.py tests/test_match_summary_from_match_state.py tests/test_transforms.py tests/test_runtime_surfaces.py
```

Related parser payload checks when game-result or game-state behavior changes:

```powershell
py -m pytest -q tests/test_gre_game_result_parser.py tests/test_gre_game_state_parser.py tests/test_parser_regressions.py
```

Full repo check before submitter opens or updates a module PR:

```powershell
.\tools\run_repo_checks.ps1
```

## Open questions

- Should underscore-prefixed extractor helpers remain the supported interface, or should a later refactor introduce public names after this audit?
- Should `_event_datetime()` continue to fall back to `datetime.now()`, or should missing timestamps be treated as an explicit empty/invalid value in a future contract?
- Should `_extract_game_result_identity()` include nested result-scope fallback behavior, or should that remain owned by `_extract_game_result_scope_result()` and `state.py`?
- Should extractor contracts distinguish top-level parser-normalized payloads from raw MTGA payloads more explicitly?

## Next Workflow Action

Next role: Module Contract Writer (B)

Pasteable prompt:

```text
Use the Mythic Edge agent constitution. Act as the Module Contract Writer thread for https://github.com/Tahjali11/Mythic-Edge/issues/9. Produce docs/contracts/parser_extractors.md based on the current src/mythic_edge_parser/app/extractors.py code, call sites, and focused tests. Distinguish observed behavior, required guarantees, unknowns, and suspected gaps. Do not implement behavior changes.
```

```yaml
workflow_handoff:
  issue: "https://github.com/Tahjali11/Mythic-Edge/issues/9"
  completed_thread: "A"
  next_thread: "B"
  source_artifact: "https://github.com/Tahjali11/Mythic-Edge/issues/9"
  target_artifact: "docs/contracts/parser_extractors.md"
  risk_tier: "High"
  branch: "codex/parser-module-audit-suite"
  validation:
    - "not run - Thinker issue only"
  stop_conditions:
    - "Do not implement behavior changes in the contract thread."
    - "Do not target module PRs at main until the parser audit suite is complete."
    - "Stop if the proposed contract would move parser-owned truth downstream."
```

**Comment by Tahjali11** [comment]
Completed via PR #10, merged into codex/parser-module-audit-suite with squash merge commit 1fe1ccaafc6c463f6defec397d5a904574720bb3. The extractors.py parser audit now has the required durable artifacts: docs/contracts/parser_extractors.md, docs/implementation_handoffs/parser_extractors_comparison.md, docs/implementation_handoffs/parser_extractors_fixer.md, and docs/contract_test_reports/parser_extractors.md. Focused extractor validation, related parser-state validation, Ruff, and PR CI passed. Closing this module audit issue; tracker #5 continues to track the broader parser audit suite.
